nnstreamer/tensor_trainer/tensor_trainer_nntrainer.hh

```diff
@@ -0,0 +1,143 @@
+ * @author Hyunil Park <hyunil46.park@samsung.com>
+ * @bug    No known bugs except for NYI items
+ */
+
+#include <condition_variable>
+#include <model.h>
+#include <nnstreamer_plugin_api.h>
+#include <nnstreamer_plugin_api_trainer.h>
+#include <vector>
+
+namespace NNTrainer {
+
+/**
+ * @brief Manage multiple inputs and labels data
+ */
+struct TensorData {
+  std::vector<char *> inputs;
+  std::vector<char *> labels;
+};
+
+class InputTensorsInfo;
+
+/**
+ * @brief NNTrainer interface for nnstreamer trainer subplugin
+ */
+class NNTrainerTrain {
+public:
+  /**
+   * @brief Construct a new NNTrainerTrain object
+   * @param prop tensor trainer subplugin properties
+   * @param _model_config model configuration file path
+   */
+  NNTrainerTrain(const GstTensorTrainerProperties *prop,
+                 const std::string &_model_config);
+
+  /**
+   * @brief Destroy the NNTrainerTrain object
+   */
+  ~NNTrainerTrain() = default;
+
+  /**
+   * @brief Create model
+   */
+  void createModel();
+
+  /**
+   * @brief Train model
+   */
+  void trainModel();
+
+  /**
+   * @brief Create dataset
+   */
+  void createDataset();
+
+  /**
+   * @brief Get NNStreamer tensor_trainer properties
+   * @param prop Tensor trainer subplugin properties
+   */
+  void getNNStreamerProperties(const GstTensorTrainerProperties *prop);
+
+  /**
+   * @brief Manage sample data
+   */
+  std::unique_ptr<NNTrainer::InputTensorsInfo> train_data, valid_data;
+  /**
+   * @brief Nntrainer dataset
+   */
+  std::shared_ptr<ml::train::Dataset> dataset_train, dataset_valid;
+
+  float training_loss, validation_loss;
+  bool is_train_complete;
+
+  int64_t tensors_inputsize[NNS_TENSOR_SIZE_LIMIT];
+  int64_t num_tensors;
+  int64_t num_inputs;
+  int64_t num_labels;
+  int64_t num_train_samples;
+  int64_t num_valid_samples;
+  int64_t total_num_samples;
+  int64_t num_epochs;
+  int64_t num_push_data;
+  std::string model_config;
+  std::string model_save_path;
+
+  GCond *train_complete_cond;
+
+private:
+  std::unique_ptr<ml::train::Model> model;
+};
+
+/**
+ * @brief Manage input tensors data and information
+ */
+class InputTensorsInfo {
+public:
+  /**
+   * @brief Construct a new InputTensorsInfo object
+   * @param _num_samples number of samples
+   * @param _num_inputs number of inputs
+   * @param _num_labels number of labels
+   * @param _tensors_inputsize[] input tensors size
+   */
+  InputTensorsInfo(int64_t _num_samples, int64_t _num_inputs,
+                   int64_t _num_labels, int64_t _tensors_inputsize[]);
+
+  /**
+   * @brief Destroy the InputTensorsInfo object
+   */
+  ~InputTensorsInfo();
+
+  bool is_data_wait_locked;
+  bool is_data_full_locked;
+  unsigned int queue_size;
+  unsigned int queue_front;
+  unsigned int queue_rear;
+  unsigned int queue_count;
+  int64_t push_count;
+  int64_t pop_count;
+  int64_t input_size[NNS_TENSOR_SIZE_LIMIT]; // feature size * data type
+  int64_t label_size[NNS_TENSOR_SIZE_LIMIT];
+  int64_t num_samples;
```

Comment:
I do not quite understand what the num_sample means. Could you add more explanation?

Reply:
"num_samples" is the total number of data used for 1 epoch during training. 
In order to set true to  "last"  of nntrainer's callback function in the subplugin, 
```
int getSample(float **input, float **label, bool *last, void *user_data)
```
need to check if the number of data coming into the subplugin is equal to num_samples.

Reply:
Overall, I will check and correct the names and descriptions of all variables.